src/Nethermind/Nethermind.Trie/PatriciaTree.cs

```diff
@@ -138,64 +138,60 @@ public void Commit(long blockNumber, bool skipRoot = false, WriteFlags writeFlag
-                while (TryDequeueCommit(out NodeCommitInfo node))
+                maxLevelForConcurrentCommit++; // Ok, we separate at top level
+                if (_writeBeforeCommit / 16 > 0)
                 {
-                    if (_logger.IsTrace) Trace(blockNumber, node);
-                    TrieStore.CommitNode(blockNumber, node, writeFlags: writeFlags);
+                    maxLevelForConcurrentCommit++; // Another level
                 }
-
-                // reset objects
-                TreePath path = TreePath.Empty;
-                RootRef!.ResolveKey(TrieStore, ref path, true, bufferPool: _bufferPool);
-                SetRootHash(RootRef.Keccak!, true);
             }
+            _writeBeforeCommit = 0;
 
-            TrieStore.FinishBlockCommit(TrieType, blockNumber, RootRef, writeFlags);
-
-            if (_logger.IsDebug) Debug(blockNumber);
-
-            bool TryDequeueCommit(out NodeCommitInfo value)
+            using (ICommitter committer = TrieStore.BeginCommit(TrieType, blockNumber, RootRef, writeFlags))
             {
-                Unsafe.SkipInit(out value);
-                return _currentCommit?.TryDequeue(out value) ?? false;
-            }
+                if (RootRef is not null && RootRef.IsDirty)
+                {
+                    TreePath path = TreePath.Empty;
+                    Commit(committer, ref path, new NodeCommitInfo(RootRef), skipSelf: skipRoot, maxLevelForConcurrentCommit: maxLevelForConcurrentCommit);
 
-            [MethodImpl(MethodImplOptions.NoInlining)]
-            void Trace(long blockNumber, in NodeCommitInfo node)
-            {
-                _logger.Trace($"Committing {node} in {blockNumber}");
+                    // reset objects
+                    RootRef!.ResolveKey(TrieStore, ref path, true, bufferPool: _bufferPool);
+                    SetRootHash(RootRef.Keccak!, true);
+                }
             }
 
+            if (_logger.IsDebug) Debug(blockNumber);
+
             [MethodImpl(MethodImplOptions.NoInlining)]
             void Debug(long blockNumber)
             {
                 _logger.Debug($"Finished committing {RootRef?.Keccak} in block {blockNumber}");
             }
         }
 
-        private void Commit(NodeCommitInfo nodeCommitInfo, bool skipSelf = false)
+        private void Commit(ICommitter committer, ref TreePath path, NodeCommitInfo nodeCommitInfo, int maxLevelForConcurrentCommit, bool skipSelf = false)
         {
             if (!_allowCommits)
             {
                 ThrowReadOnlyTrieException();
             }
 
             TrieNode node = nodeCommitInfo.Node;
-            TreePath path = nodeCommitInfo.Path;
             if (node!.IsBranch)
             {
-                // idea from EthereumJ - testing parallel branches
-                if (!_parallelBranches || !nodeCommitInfo.IsRoot)
+                if (path.Length > maxLevelForConcurrentCommit)
```

Comment:
So here instead of checking path.Length, wouldn't be better to check if we have a free thread to take (and maybe that path is at least 3-4)?

Reply:
Its in the `committer.CanSpawnTask()` part that finally decide if there are threads available. This is limiting it to top level in case there arent enough node to justify multithreading at all.